cyto_dl/models/im2im/diffusion_autoencoder.py

```diff
@@ -0,0 +1,227 @@
+                save_intermediates=True,
+                conditioning=cond,
+                verbose=False,
+                intermediate_steps=1,
+            )
+        # during training, final image is all-nan, try to return last non-nan image
+        if torch.any(torch.isnan(sample)):
+            while torch.any(torch.isnan(sample)) and len(intermediates) > 0:
+                sample = intermediates.pop(-1)
+        return sample
+
+    def generate_image(self, stage, cond_img, diff_img):
+        """Save the sequence of denoising steps."""
+        self.scheduler.set_timesteps(num_inference_steps=100)
+        with torch.no_grad():
+            cond = self.semantic_encoder(cond_img).unsqueeze(2)
+        noise = torch.randn_like(cond_img, device=self.device)
+        sample = self._generate_image(noise, cond)
+
+        for img, name in zip([cond_img, diff_img, sample], ["cond", "diff", "recon"]):
+            OmeTiffWriter.save(
+                uri=f"{self.hparams.save_dir}/{self.trainer.current_epoch}_{stage}_{name}.tiff",
+                data=detach(img).astype(float),
+            )
+
+    def _to_tensor(self, batch):
+        """convert monai metatensors to tensors."""
+        for k, v in batch.items():
+            if isinstance(v, MetaTensor):
+                batch[k] = v.as_tensor()
+        return batch
+
+    def model_step(self, stage, batch, batch_idx):
+        batch = self._to_tensor(batch)
+        cond_img = batch[self.hparams.condition_key]
+        diff_img = batch[self.diffusion_key]
+        noise, noise_pred, latent = self.forward(cond_img, diff_img)
+        if (self.trainer.current_epoch % self.hparams.save_images_every_n_epochs) == batch_idx == 0:
+            self.generate_image(stage, cond_img[:1], diff_img[:1])
+        diffusion_loss = self.loss(noise, noise_pred)
+
+        return {"loss": diffusion_loss}, latent, None
+
+    def validation_step(self, batch, batch_idx):
+        loss, preds, targets = self.model_step("val", batch, batch_idx)
+        self.compute_metrics(loss, preds, targets, "val")
+        return loss, preds
+
+    def latent_walk(self, cond, save_name):
+        self.scheduler.set_timesteps(num_inference_steps=100)
+        with torch.no_grad():
+            recon = None
+            for _ in tqdm.tqdm(range(int(self.hparams.n_noise_samples)), desc='Sampling'):
+                # keep noise constant across walk for consistency
+                noise = torch.stack([torch.randn([1] + self.hparams.image_shape, device=self.device)] * cond.shape[0])
+                sample = self._generate_image(noise, cond.unsqueeze(2))
+                recon = sample if recon is None else recon + sample
+            # average reconstruction across samples
+            recon /= self.hparams.n_noise_samples
+        OmeTiffWriter.save(
+            uri=f"{self.hparams.save_dir}/{save_name}.tiff", data=detach(recon).astype(float)
+        )
+
+    def _generate_and_re_embed(self, z, batch_idx):
```

Comment:
cool!! is this loss not a part of training as well?

Reply:
This was more a validation to see if the semantic latent was actually being used, maybe not worth including in the PR.